### DISCUSSION TITLE
concurrency confusion around `OnRealtimeBeforeSubscribeRequest`

### DISCUSSION THREAD

**Discussion by dotmrjosh** [pocketbase/pocketbase#5198] [Q&A] [answered]
## Actual problem
The issue was actually my own fault in dealing with concurrency and not the originally proposed question. Below is an example of the problematic code:

```go
package main

import (
  "fmt"
  "time"

  "github.com/pocketbase/pocketbase"
  "github.com/pocketbase/pocketbase/core"
)

func main() {
  app := pocketbase.New()

  app.OnRealtimeBeforeSubscribeRequest().Add(func(e *core.RealtimeSubscribeEvent) error {
    e.Client.Set("key", "value")
    println("subscription done")
    return nil
  })

  // Some other code at a later event
  done := make(chan struct{})
  go func() {
  Loop:
    for {
      select {
      case <-done:
        break Loop
      case <-time.After(100 * time.Millisecond):
        for _, client := range app.SubscriptionsBroker().Clients() {
          // Problems lie here
          fmt.Printf("%s", client.Get("key").(string))
        }
      }
    }
  }()

  app.OnTerminate().Add(func(e *core.TerminateEvent) error {
    close(done)
    return nil
  })

  if err := app.Start(); err != nil {
    println("Failed to start app")
  }
}
```

## Answer

In the goroutine which prints messages on intervals, there was no check to ensure that the subscription hook had completed (since Connecting and Subscription are 2 separate things, I was originally under the wrong assumption that subscription *was* the client connecting). The fix is simply adding following check where the problem comment is:

```go
if !client.HasSubscription("your-subscription") {
  continue
}
```

<details>
<summary>Original question</summary>
<hr>

I'm working on a hook that does some specific validation before a user is allowed to subscribe to some realtime event.
If I store a value using `e.Client.Set("key", "value")`, the value is not accessible from the apps SubscriptionsBroker().Clients():

## Example:

```go
func SomeCustomThings(app *pocketbase.PocketBase) {
  app.OnRealtimeBeforeSubscribeRequest().Add(func(e *core.RealtimeSubscribeEvent) error {
    e.Client.Set("key", "value")
  }

  // Some other code at a later event
  go func() {
    for _, client := range app.SubscrptionsBroker().Clients() {
      println(client.Get("key")) // Fails!
    }
  }()
}
```

## Workaround

The current workaround I have is just to always use the apps SubscriptionsBroker with ClientById but unsure if this is the intended method for this:

```go
func SomeCustomThings(app *pocketbase.PocketBase) {
  app.OnRealtimeBeforeSubscribeRequest().Add(func(e *core.RealtimeSubscribeEvent) error {
    client, err := app.SubscrptionsBroker().ClientById(e.Client.Id)
    if err != nil {
      return err
    }
    client.Set("key", "value")
  }

  // Some other code at a later event
  go func() {
    for _, client := range app.SubscrptionsBroker().Clients() {
      println(client.Get("key")) // Works correctly now
    }
  }()
}
```

I have not tested the After variation of the hook to see if the same issue exists but would assume the same might apply.

### Versions

```
go 1.22.3
github.com/pocketbase/pocketbase v0.22.14
```
</details>

**Comment by ganigeorgiev** [comment]
The above code doesn't seem correct (maybe is missing parenthesis?) and I don't understand what is the question or the issue here.

[`SubscrptionsBroker().Clients()`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.16/tools/subscriptions#Broker.Clients) returns a shallow copy of the registered clients map but the `Client` represents an interface which is backed by `*subscriptions.DefaultClient` pointer, or in other words, both the iterated client and the one from the hook event are pointing to the same instance.


Please elaborate or provide clear reproducible steps if you think that this is a PocketBase issue.

**Reply by dotmrjosh** [reply, marked as answer]
Apologies for the hacked mess, was trying to pull out the related info from the larger project at hand. I think I found the issue where it was a concurrency and a wrong assumption of the hook.

The wrong assumption was that a client was not connected until a subscription was made. So the concurrent code would find the client before the subscription was made and try and pull values that weren't yet assigned.

Apologies for the public rubber ducking and any time wasted. Will update the post with the actual issue and answer if anyone comes across the same problem.
